src/proposal.mligo

```diff
@@ -408,6 +484,39 @@ let drop_proposal (proposal_key, config, store : proposal_key * config * storage
   else
     (failwith("FAIL_DROP_PROPOSAL_NOT_OVER") : return)
 
+let update_period_frozen_history(amt, current_period, period_frozen_history: int * nat * period_frozen_history) : period_frozen_history
+  = match is_nat(current_period - period_frozen_history.updated_period) with
```

Comment:
If picking the representation in the comment above here we should do two things (in this order).

First: check if the `quorum_threshold_at_stage.last_updated_period` is not `current_period`.
In case it isn't, we need to:
1. calculate the new `quorum_threshold`, which we can do as we have all the info from a previous period
2. set the `last_updated_period` to `current_period`

Second, regardless of what we do above we now know that we are storing the info for the `current_period`, so we can update the `frozen_supply` with the given value.
Note: BTW in this representation we can "safely" subtract in case of unfreezing: if we track the `total_supply` we know that no value should not be higher than that,so if this this is not true we should fail as there's a bug in the contract.